static/js/publisher/metrics/graphs/activeDevices.js

```diff
@@ -3,418 +3,730 @@ import { format } from "d3-format";
+
+    this.hasRendered = false;
+
+    this.lines = line()
+      .x(d => this.xScale(d.date))
+      .y(d => this.yScale(d.value));
+
+    this.areas = area()
+      .curve(curveMonotoneX)
+      .x(d => this.xScale(d.data.date))
+      .y0(d => this.yScale(d[0]))
+      .y1(d => this.yScale(d[1]));
+
+    this.shortValue = number =>
+      number < 1000 ? number : format(".2s")(number);
+    this.commaValue = number => format(",")(number);
+    this.tooltipTimeFormat = utcFormat("%Y-%m-%d");
+
+    this.prepareSVG();
+
+    const resize = debounce(() => {
+      if (this.hasRendered) {
+        this.prepareSVG()
+          .prepareData()
+          .render()
+          .enableTooltip();
+      }
+    }, 100);
 
-const shortValue = number => (number < 1000 ? number : format(".2s")(number));
-const commaValue = number => format(",")(number);
-const tooltipTimeFormat = utcFormat("%Y-%m-%d");
-
-let graphType;
-let metricsDefaultTrack;
-
-const xScaleFunc = (padding, width, data) => {
-  return scaleLinear()
-    .rangeRound([padding.left, width - padding.left - padding.right])
-    .domain(extent(data, d => d.date));
-};
-
-const yScaleFunc = (padding, height, max) => {
-  return scaleLinear()
-    .rangeRound([height - padding.top - padding.bottom, padding.top])
-    .nice()
-    .domain([0, max + Math.ceil(max * 0.1)]);
-};
-
-const colourScaleFunc = keys => {
-  return scaleOrdinal(schemePaired).domain(keys);
-};
-
-const getStackedData = (keys, data) => {
-  const stackFunc = stack()
-    .order(stackOrderReverse)
-    .keys(keys);
-
-  return stackFunc(data);
-};
-
-const bisectDate = bisector(d => d.date).left;
-
-const tooltipRows = (data, currentHoverKey, colorScale) => {
-  let dataArr = [];
-  let other = {
-    key: "other",
-    value: 0,
-    count: 0
-  };
-
-  let keys = Object.keys(data).filter(key => key !== "date");
-
-  if (graphType === "channel") {
-    keys = sortChannels(keys, {
-      defaultTrack: metricsDefaultTrack
-    }).list;
+    select(window).on("resize", resize);
   }
 
-  keys.forEach(key => {
-    dataArr.push({
-      key: key,
-      value: data[key]
-    });
-  });
+  colorScaleGenerator() {
+    return scaleOrdinal(schemePaired).domain(this.keys);
+  }
 
-  if (graphType !== "channel") {
-    dataArr.sort((a, b) => b.value - a.value);
+  xScaleGenerator() {
+    return scaleLinear()
+      .rangeRound([
+        this.padding.left,
+        this.width - this.padding.left - this.padding.right
+      ])
+      .domain(extent(this.data, d => d.date));
   }
 
-  if (dataArr.length > 15) {
-    dataArr.forEach((item, index) => {
-      if (index >= 15) {
-        other.value += item.value;
-        other.count += 1;
-        other.key = `${other.count} other`;
-        item.skip = true;
-      }
-    });
+  yScaleGenerator() {
+    return scaleLinear()
+      .rangeRound([
+        this.height - this.padding.top - this.padding.bottom,
+        this.padding.top
+      ])
+      .nice()
```

Comment:
Why `xScaleGenerator` is not `.nice`? - me having no idea what `nice` does...

Reply:
"Extends the domain so that it starts and ends on nice round values." :D!
I can't find `notNice` in the docs :(